From 6a5de63a6b6f3a22c3036f9f559cebb7342eb563 Mon Sep 17 00:00:00 2001
From: Murat Sezgin <msezgin@codeaurora.org>
Date: Tue, 27 Jan 2015 11:08:27 -0800
Subject: [PATCH] ipv6: Add api to get dev for ipv6 addr.

API function takes an IPv6 address and returns
a hold to a net device that has a matching address.

Change-Id: I21af606193fafd68a61249e893749b87f6b36c76
Signed-off-by: Murat Sezgin <msezgin@codeaurora.org>
---
 include/net/addrconf.h |  2 ++
 net/ipv6/addrconf.c    | 29 +++++++++++++++++++++++++++++
 2 files changed, 31 insertions(+)

--- a/include/net/addrconf.h
+++ b/include/net/addrconf.h
@@ -109,6 +109,8 @@ int ipv6_chk_prefix(const struct in6_add
 
 struct net_device *ipv6_dev_find(struct net *net, const struct in6_addr *addr,
 				 struct net_device *dev);
+struct net_device *ipv6_dev_find_qca(struct net *net, struct in6_addr *addr,
+				 int strict);
 
 struct inet6_ifaddr *ipv6_get_ifaddr(struct net *net,
 				     const struct in6_addr *addr,
--- a/net/ipv6/addrconf.c
+++ b/net/ipv6/addrconf.c
@@ -1003,6 +1003,7 @@ void inet6_ifa_finish_destroy(struct ine
 
 	kfree_rcu(ifp, rcu);
 }
+EXPORT_SYMBOL(inet6_ifa_finish_destroy);
 
 static void
 ipv6_link_dev_addr(struct inet6_dev *idev, struct inet6_ifaddr *ifp)
@@ -2065,6 +2066,36 @@ struct inet6_ifaddr *ipv6_get_ifaddr(str
 
 	return result;
 }
+EXPORT_SYMBOL(ipv6_get_ifaddr);
+
+/* ipv6_dev_find_qca()
+ *	Find (and hold) net device that has the given address.
+ *	Or NULL on failure.
+ */
+struct net_device *ipv6_dev_find_qca(struct net *net, struct in6_addr *addr,
+				 int strict)
+{
+	struct inet6_ifaddr *ifp;
+	struct net_device *dev;
+
+	ifp = ipv6_get_ifaddr(net, addr, NULL, strict);
+	if (!ifp)
+		return NULL;
+
+	if (!ifp->idev) {
+		in6_ifa_put(ifp);
+		return NULL;
+	}
+
+	dev = ifp->idev->dev;
+	if (dev)
+		dev_hold(dev);
+
+	in6_ifa_put(ifp);
+
+	return dev;
+}
+EXPORT_SYMBOL(ipv6_dev_find_qca);
 
 /* Gets referenced address, destroys ifaddr */
 
